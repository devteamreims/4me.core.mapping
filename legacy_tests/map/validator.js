import {APPDIR} from '../settings'
// Stubs
import * as CwpTreeStub from '../stubs/cwp/Tree';
import * as SectorTreeStub from '../stubs/sector/Tree';

const modulePath = APPDIR + 'mapping/validator';
const moduleStubs = {};

let cwpTree = new CwpTreeStub.default();
let sectorTree = new SectorTreeStub.default();

describe('mapValidator', () => {
  it('should load', () => {
    let mapValidator = proxyquire(modulePath, moduleStubs).default;
  });

  describe('validator', () => {
    let validator;
    beforeEach(() => {
      validator = proxyquire(modulePath, moduleStubs).default;
    });

    describe('validate', () => {
      it('should have a validate method', () => {
        validator.validate.should.be.a('function');
      });

      it('should refuse invalid input', () => {
        expect(() => validator.validate('string', cwpTree, sectorTree)).to.throws(/invalid argument/i);
        expect(() => validator.validate(null, cwpTree, sectorTree)).to.throws(/invalid argument/i);
      });

      it('should refuse invalid format', () => {
        let invalid = [
          {
            cwpId: 2,
            foo: 'bar'
          }, {
            foo: 'bar'
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/format/i);
      });

      it('should refuse unknown CWPs', () => {
        let invalid = [
          {
            cwpId: 65,
            sectors: ['UF', 'KF', 'KD']
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/unknown cwp/i);
      });

      it('should refuse CWPs with the wrong type', () => {
        // See test/stubs/cwpStaticData.js
        let invalid = [
          {
            cwpId: 1,
            sectors: ['UF', 'KF', 'KD']
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/wrong type/i);
      });

      it('should refuse disabled CWPs', () => {
        // See test/stubs/cwpStaticData.js
        let invalid = [
          {
            cwpId: 6,
            sectors: ['UF', 'KF', 'KD']
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/disabled/i);
      });

      it('should refuse unknown sectors', () => {
        let invalid = [
          {
            cwpId: 4,
            sectors: ['UF', 'KF', 'KD', '4M']
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/unknown sector/i);
      });

      it('should refuse cwps mentionned twice', () => {
        let invalid = [
          {
            cwpId: 4,
            sectors: ['KF', 'KD']
          }, {
            cwpId: 4,
            sectors: ['UF']
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/multiple times/i);
      });

      it('should refuse sectors bound twice', () => {
        let invalid = [
          {
            cwpId: 4,
            sectors: ['UF', 'KF', 'KD']
          }, {
            cwpId: 3,
            sectors: ['UF']
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/multiple times/i);
      });

      it('should reject if a sector is missing', () => {
        let invalid = [
          {
            cwpId: 4,
            sectors: ['UF', 'KF']
          }
        ];
        expect(() => validator.validate(invalid, cwpTree, sectorTree)).to.throws(/missing/i);
      });

      it('should return true provided a valid map', () => {
        let valid = [
          {
            cwpId: 4,
            sectors: ['UF', 'KF']
          }, {
            cwpId: 3,
            sectors: ['KD']
          }
        ];

        expect(() => validator.validate(valid, cwpTree, sectorTree)).to.not.throws();
        validator.validate(valid, cwpTree, sectorTree).should.eql(true);
      });

      it('should accept disabled positions without sectors bound', () => {
        let valid = [
          {
            cwpId: 4,
            sectors: ['UF', 'KF']
          }, {
            cwpId: 3,
            sectors: ['KD']
          }, {
            cwpId: 6,
            sectors: []
          }
        ];

        expect(() => validator.validate(valid, cwpTree, sectorTree)).to.not.throws();
        validator.validate(valid, cwpTree, sectorTree).should.eql(true);
      });

      it('should accept other position types without sectors bound', () => {
        let valid = [
          {
            cwpId: 4,
            sectors: ['UF', 'KF']
          }, {
            cwpId: 3,
            sectors: ['KD']
          }, {
            cwpId: 1,
            sectors: []
          }
        ];

        expect(() => validator.validate(valid, cwpTree, sectorTree)).to.not.throws();
        validator.validate(valid, cwpTree, sectorTree).should.eql(true);
      });
    });
  });
});