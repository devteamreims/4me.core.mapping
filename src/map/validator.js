import _ from 'lodash';
import R from 'ramda';
import d from 'debug';

const debug = d('4me.mapping.validator');

import { clients, sectors } from '../env';

  // Validates a map
export function validate(map) {
  // Input sanitation
  if(!_.isArray(map)) {
    throw new Error('Invalid argument');
  }

  // Validate format of each element in the array
  let mentionnedClientIds = [];
  let mentionnedSectors = [];
  map.forEach((m) => {
    const isValid = m.cwpId && (_.isArray(m.sectors) || !!m.disabled);
    if(!isValid) {
      throw new Error('Invalid argument : wrong format');
    }

    // Check if every CWP exists, is the right type and is not disabled
    const client = clients.getClientById(m.cwpId);

    if(!client) {
      throw new Error(`Trying to assign sectors to an unknown CWP (#${m.cwpId})`);
    }

    if(client.type !== 'cwp') {
      throw new Error(`Map should not reference clients with a non-CWP type (#${m.cwpId} / ${client.type})`);
    }

    if(m.disabled === true && m.sectors && m.sectors.length > 0) {
      throw new Error(`Trying to assign sectors to a disabled CWP (#${m.cwpId})`);
    }

    // Check for duplicate declarations of a specific CWP
    if(mentionnedClientIds.includes(m.cwpId)) {
      throw new Error(`Map should not reference a client multiple times (#${m.cwpId})`);
    }

    mentionnedClientIds.push(m.cwpId);

    // If we don't have sectors set for this particular item, break the loop and move to the next item
    if(!m.sectors) {
      return;
    }

    const areInEnvironment = R.all(R.contains(R.__, sectors.getElementarySectors()));
    if(!areInEnvironment(m.sectors)) {
      throw new Error(`Trying to bind unknown sectors (#${m.cwpId})`);
    }

    const containsDuplicateSectors = R.pipe(
      R.intersection(mentionnedSectors),
      R.isEmpty,
      R.not,
    );

    if(containsDuplicateSectors(m.sectors)) {
      throw new Error(`Trying to bind sectors multiple times (#${m.cwpId})`);
    }

    mentionnedSectors = mentionnedSectors.concat(m.sectors);

  });

  // Finally we check we have all elementary sectors bound somewhere
  // :: (mentionedSectors, allSectors) => missingSectors
  const getMissingSectors = R.without;
  // :: (mentionnedSectors, allSectors) => boolean
  const containsMissingSectors = R.pipe(
    getMissingSectors,
    R.isEmpty,
    R.not,
  );

  // Check if every sector is bound
  if(containsMissingSectors(mentionnedSectors, sectors.getElementarySectors())) {
    const missingSectors = getMissingSectors(mentionnedSectors, sectors.getElementarySectors());
    throw new Error(`Sectors are missing from the map (${missingSectors.join(',')})`);
  }

  return true;
}

export default validate;
