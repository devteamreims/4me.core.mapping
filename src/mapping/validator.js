import _ from 'lodash';
import d from 'debug';
import Promise from 'bluebird';
const debug = d('4me.mapping.validator');


  // Validates a map
export function validate(map, cwpTree, sectorTree) {
  // Input sanitation
  if(!_.isArray(map)) {
    throw new Error('Invalid argument');
  }

  // Validate format of each element in the array
  map.forEach((m) => {
    if(!_.isNumber(m.cwpId) || !_.isArray(m.sectors)) {
      throw new Error('Invalid argument : wrong format');
    }
  });

  // Check if every CWP exists, is the right type and is not disabled
  map.forEach((m) => {
    let cwp = cwpTree.getById(m.cwpId);

    if(!cwp) {
      throw new Error(`Trying to assign sectors to an unknown CWP (#${m.cwpId})`);
    }

    // No sector assigned to this CWP
    if(_.isEmpty(m.sectors)) {
      return;
    }

    if(cwp.disabled === true) {
      throw new Error(`Trying to assign sectors to a disabled CWP (#${m.cwpId})`);
    }
    if(cwp.isCwp() !== true) {
      throw new Error(`Trying to assign sectors to a CWP with the wrong type (#${m.cwpId} / ${cwp.type})`);
    }
  });

  // Check for CWPs mentionned twice or more
  let cwpIds = _.map(map, (m) => m.cwpId);
  if(!_.eq(cwpIds, _.uniq(cwpIds))) {
    throw new Error(`Trying to bind sectors to a CWP multiple times`);
  }

  let boundSectors = _.flatten(map.map((m) => m.sectors)).map((s) => s.toUpperCase());
  // Check for duplicate sectors
  if(_.uniq(boundSectors).length !== boundSectors.length) {
    throw new Error(`Trying to bind sectors multiple times`);
  }
  // Check if every sector exists
  let elementarySectors = sectorTree.getElementary().map((s) => s.name.toUpperCase());

  if(_.difference(boundSectors, elementarySectors).length !== 0) {
    throw new Error('Trying to bind unknown sectors');
  }

  // Check if every sector is bound
  if(_.difference(elementarySectors, boundSectors).length !== 0) {
    throw new Error('We have missing elementary sectors !');
  }

  return true;
}

export default {
  validate: validate
};