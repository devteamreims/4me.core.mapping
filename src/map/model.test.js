import {
  generateMap,
  set,
  get,
} from './model';
import { validate } from './validator';

describe('map::model', () => {
  describe('generateMap', () => {
    test('should generate a map', () => {
      const map = generateMap();
      expect(map).toMatchSnapshot();
    });

    test('should generate a valid map', () => {
      const map = generateMap();
      expect(validate(map)).toBe(true);
    });
  });

  describe('set', () => {
    test('should validate map', () => {
      const invalidMap = [{cwpId: 11, disabled: true}];
      return set(invalidMap).then(
        () => expect(true).toBe(false),
        err => expect(err).toMatchSnapshot(),
      );
    });

    const newMap = [{
      cwpId: 11,
      sectors: [
        "UR",
        "XR",
        "KR",
        "HYR",
        "UB",
        "UN",
        "KN",
        "HN",
        "UE",
        "XE",
        "KE",
        "HE",
        "UH",
        "XH",
        "KH",
        "HH",
        "KD",
        "UF",
        "KF",
      ],
    }, {
      cwpId: 12,
      sectors: ['E', 'SE'],
    }];

    test('should resolve to the new map', () => {
      return set(newMap).then(
        map => expect(map).toEqual(newMap),
      );
    });

    test('should properly set the map in the database', () => {
      return set(newMap)
        .then(() => get())
        .then(map => expect(map).toEqual(newMap));
    });
  });
});
