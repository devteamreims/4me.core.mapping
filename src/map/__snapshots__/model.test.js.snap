exports[`map::model generateMap should generate a map 1`] = `
Array [
  Object {
    "cwpId": 11,
    "sectors": Array [
      "UR",
      "XR",
      "KR",
      "HYR",
      "UB",
      "UN",
      "KN",
      "HN",
      "UE",
      "XE",
      "KE",
      "HE",
      "UH",
      "XH",
      "KH",
      "HH",
      "KD",
      "UF",
      "KF",
      "E",
      "SE",
    ],
  },
]
`;

exports[`map::model set should validate map 1`] = `[Error: Sectors are missing from the map (UR,XR,KR,HYR,UB,UN,KN,HN,UE,XE,KE,HE,UH,XH,KH,HH,KD,UF,KF,E,SE)]`;
